gdal/ogr/ogrsf_frmts/flatgeobuf/ogrflatgeobuflayer.cpp

```diff
@@ -0,0 +1,829 @@
+    if (crs != nullptr) {
+        m_poSRS = new OGRSpatialReference();
+        auto org = crs->org();
+        auto code = crs->code();
+        auto definition = crs->definition();
+        auto definition_12_063 = crs->definition_12_063();
+        m_poSRS->SetAuthority(nullptr, org->c_str(), code);
+        if (org == nullptr && code != 0) {
+            m_poSRS->importFromEPSG(code);
+        } else if (definition_12_063) {
+            m_poSRS->importFromWkt(definition_12_063->c_str());
+        } else if (definition) {
+            m_poSRS->importFromWkt(definition->c_str());
+        }
+    }
+
+    auto eGType = getOGRwkbGeometryType();
+
+    const char *pszName = nullptr;
+    if (m_poHeader->name()) {
+        pszName = m_poHeader->name()->c_str();
+        CPLDebug("FlatGeobuf", "header name: %s", pszName);
+    }
+    m_poFeatureDefn = new OGRFeatureDefn(pszName);
+    SetDescription(m_poFeatureDefn->GetName());
+    m_poFeatureDefn->SetGeomType(wkbNone);
+    OGRGeomFieldDefn *poGeomFieldDefn = new OGRGeomFieldDefn(nullptr, eGType);
+    if (m_poSRS != nullptr)
+        poGeomFieldDefn->SetSpatialRef(m_poSRS);
+    m_poFeatureDefn->AddGeomFieldDefn(poGeomFieldDefn, false);
+    readColumns();
+    m_poFeatureDefn->Reference();
+}
+
+OGRFlatGeobufLayer::OGRFlatGeobufLayer(
+    const char *pszLayerName,
+    const char *pszFilename,
+    OGRSpatialReference *poSpatialRef,
+    OGRwkbGeometryType eGType)
+{
+    CPLDebug("FlatGeobuf", "Request to create layer %s", pszLayerName);
+
+    // TODO: free
+    m_pszLayerName = CPLStrdup(pszLayerName);
+    m_pszFilename = CPLStrdup(pszFilename);
+    m_create = true;
+    m_eGType = eGType;
+    translateOGRwkbGeometryType();
+    if (poSpatialRef)
+        m_poSRS = poSpatialRef->Clone();
+
+    CPLDebug("FlatGeobuf", "eGType: %d", (int) eGType);
+    CPLDebug("FlatGeobuf", "m_geometryType: %d", (int) m_geometryType);
+
+    SetMetadataItem(OLMD_FID64, "YES");
+
+    m_poFeatureDefn = new OGRFeatureDefn(pszLayerName);
+    SetDescription(m_poFeatureDefn->GetName());
+    m_poFeatureDefn->SetGeomType(eGType);
+    m_poFeatureDefn->Reference();
+}
+
+void OGRFlatGeobufLayer::translateOGRwkbGeometryType()
+{
+    switch (wkbFlatten(m_eGType)) {
+        case OGRwkbGeometryType::wkbPoint: m_geometryType = GeometryType::Point; break;
+        case OGRwkbGeometryType::wkbMultiPoint: m_geometryType = GeometryType::MultiPoint; break;
+        case OGRwkbGeometryType::wkbLineString: m_geometryType = GeometryType::LineString; break;
+        case OGRwkbGeometryType::wkbMultiLineString: m_geometryType = GeometryType::MultiLineString; break;
+        case OGRwkbGeometryType::wkbPolygon: m_geometryType = GeometryType::Polygon; break;
+        case OGRwkbGeometryType::wkbMultiPolygon: m_geometryType = GeometryType::MultiPolygon; break;
+        default:
+            CPLError(CE_Fatal, CPLE_NotSupported, "toGeometryType: Unknown OGRwkbGeometryType %d", (int) m_eGType);
```

Comment:
CE_Fatal should not be used here or anywhere else (other instances below). It crashes the process

Reply:
Switched to CE_Failure